lib/mkCargoDerivation.nix

```diff
@@ -39,13 +40,55 @@ let
     "pnameSuffix"
     "stdenv"
   ];
+
+  rustcWrapper = writeShellApplication {
+    name = "rustc-wrapper";
+    text = ''
+      set -euo pipefail
+
+      args=("$@")
+
+      temp_dir=$(mktemp -d)
+      trap 'rm -rf -- "$temp_dir"' EXIT
+
+      for i in "''${!args[@]}"; do
+        if [[ "''${args[i]}" == --out-dir=* ]]; then
+          current_out_dir="''${args[i]#--out-dir=}"
+          args[i]="--out-dir=$temp_dir"
+          break
+        elif [[ "''${args[i]}" == --out-dir ]]; then
+          current_out_dir="''${args[i+1]}"
+          args[i+1]="$temp_dir"
+          break
+        fi
+      done
+
+      if [[ -v current_out_dir ]]; then
+        stderr="$temp_dir/crane_stderr"
+        set +e
+        stdout=$("''${args[@]}" 2>"$stderr")
+        exit_code=$?
+        set -e
+        stderr_text=$(cat "$stderr")
```

Comment:
Just a thought - if you know out dir ... wouldn't it be faster & easier to delete it beforehand instead of changing it and copying from it? Technically cargo might want to read something from it first... but something tells me it never actually happen.

Reply:
I would also be a bit worries about these temp paths being written somewhere and then causing rebuilds etc. Instead of making them random, maybe they should be fixed. Eg. always `$TMPDIR/rustc-wrapper/${orig-outdir-absolute-path}`.

Reply:
Yeah, I considered this. The problem is that what rustc outputs is dependent on what rustc has been asked to do (tell it to make a dylib? it'll make one!), so it's much more tightly coupled. Whereas the output dir is loosely coupled.

Regarding the temp paths being written somewhere - I logicked myself that this should never happen. In particular, compilers put a lot of energy into being reproducible - it's dodgy if the output is dependent on the path the output is to be written to.